ksqldb-functional-tests/src/test/resources/query-validation-tests/fk-join.json

```diff
@@ -53,131 +53,557 @@
     },
     {
-      "name": "Should fail on outer-join",
+      "enabled": false,
+      "name": "value column - inner join",
       "properties": {
         "ksql.joins.foreign.key.enable": true
       },
       "statements": [
-        "CREATE TABLE left_table (id1 BIGINT PRIMARY KEY, f1 BIGINT) WITH (kafka_topic='left_topic', format='JSON');",
-        "CREATE TABLE right_table (id2 BIGINT PRIMARY KEY, f2 BIGINT) WITH (kafka_topic='right_topic', format='JSON');",
-        "CREATE TABLE output AS SELECT id1, id2, f1, f2 FROM left_table FULL OUTER JOIN right_table ON f1 = id2;"
+        "CREATE TABLE left_table (L_ID BIGINT PRIMARY KEY, NAME varchar, VALUE bigint) WITH (kafka_topic='left_topic', value_format='JSON');",
+        "CREATE TABLE right_table (R_ID BIGINT PRIMARY KEY, F1 varchar, F2 bigint) WITH (kafka_topic='right_topic', value_format='JSON');",
+        "CREATE TABLE OUTPUT as SELECT l_id, as_value(r_id) as r_id, name, f1 FROM left_table JOIN right_table ON value = r_id;"
+      ],
+      "inputs": [
+        {"topic": "right_topic", "key": 0, "value": {"F1": "blah", "F2": 4}, "timestamp": 0},
+        {"topic": "left_topic", "key": 1, "value": {"NAME": "zero", "VALUE": 0}, "timestamp": 10000},
+        {"topic": "left_topic", "key": 0, "value": {"NAME": "foo", "VALUE": 100}, "timestamp": 13000},
+        {"topic": "right_topic", "key": 0, "value": {"F1": "a", "F2": 10}, "timestamp": 15000},
+        {"topic": "left_topic", "key": 10, "value": {"NAME": "bar", "VALUE": 0}, "timestamp": 16000},
+        {"topic": "right_topic", "key": 0, "value": null, "timestamp": 17000},
+        {"topic": "left_topic", "key": 1, "value": null, "timestamp": 18000}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "key": 1, "value": {"R_ID": 0, "NAME": "zero", "F1": "blah"}, "timestamp": 10000},
+        {"topic": "OUTPUT", "key": 1, "value": {"R_ID": 0, "NAME": "zero", "F1": "a"}, "timestamp": 15000},
+        {"topic": "OUTPUT", "key": 10, "value": {"R_ID": 0, "NAME": "bar", "F1": "a"}, "timestamp": 16000},
+        {"topic": "OUTPUT", "key": 1, "value": null, "timestamp": 17000},
+        {"topic": "OUTPUT", "key": 10, "value": null, "timestamp": 17000}
+      ],
+      "post": {
+        "sources": [
+          {"name": "OUTPUT", "type": "table", "schema": "L_ID BIGINT KEY, `R_ID` BIGINT, `NAME` STRING, `F1` STRING"}
+        ]
+      }
+    },
+    {
+      "enabled": false,
+      "name": "value column - left join",
+      "properties": {
+        "ksql.joins.foreign.key.enable": true
+      },
+      "statements": [
+        "CREATE TABLE left_table (L_ID BIGINT PRIMARY KEY, NAME varchar, VALUE bigint) WITH (kafka_topic='left_topic', value_format='JSON');",
+        "CREATE TABLE right_table (R_ID BIGINT PRIMARY KEY, F1 varchar, F2 bigint) WITH (kafka_topic='right_topic', value_format='JSON');",
+        "CREATE TABLE OUTPUT as SELECT l_id, as_value(r_id) as r_id, name, f1 FROM left_table LEFT JOIN right_table ON value = r_id;"
```

Comment:
This is just a random thought while reading this test, and is not really for the PR itself: I feel that now with PK-join and FK-join together, the current PK "inheritance" semantics is a bit odd. More specifically, with a CTAS statement from a join, seems our scenarios are:

1) `SELECT left.key, right.key` will throw an error from either PK and FK, however conceptually for PK-join it is okay to have both, which would always be the same.

2) `SELECT *` will only set `left.key` as the key for the resulted table, and the `right.key` would be moved to values.

3) If user do want to have both `left.key` and `right.key` they need to call this `as_value()` function to cast the `right.key` to a value. This also means that for FK-join, the foreign-key must be on the left-hand side, and ksql would not be able to auto-switch the left/right if we have `left JOIN right on left.key = right.fk-value`.

I'm feeling that we should try to be more flexible about inheriting the key columns from a join statement, instead of enforcing users to make `as_value()` calls. Say in the above case, would it be better if:

1) `SELECT left.key, right.key` would be allowed, and we would look at the join condition to determine if the `left.key`, or `right.key` (for FK-join) or both together (for PK-join) would be the new table's PK.

2) `SELECT *` would similarly depend on the join conditions, to set either `left.key`, or `right.key` or both together as the new table's PK.

Reply:
Hi @guozhangwang , thanks for raising this. I played around with existing table-table joins some more and discovered some discrepancies in my understanding of how ksqlDB handles keys in join projections, your understanding, and what actually happens. Let me start with some background on my understanding, with the caveat that some of my understanding turns out to be incorrect.

ksqlDB currently requires that users select all key columns in the projection of their persistent queries, across the board. If a key column is left out, they will receive an error. If a key column is duplicated (e.g., `select k, k as key_alias, ...`) they will receive an error, since this indicates that they are trying to change the key column(s), which is not allowed. If they want a key column to appear in the result value, they can use `AS_VALUE()` to achieve this. If a user truly wishes to change the key column(s) of the output, they can use `PARTITION BY`. There has been [discussion](https://github.com/confluentinc/ksql/issues/6372) around lifting these stringent requirements for having keys appear exactly as expected in the `SELECT` clause of the query, but it has thus far been inconclusive.

If you extend these principles to joins, then key selection for primary key (non-outer) joins should have the following semantics. Suppose the join is `left JOIN right ON left.k = right.k`:
* If the projection includes `left.k` but not `right.k`, this should succeed.
* If the projection includes `right.k` but not `left.k`, this should succeed.
* If the projection includes both `left.k` and `right.k`, then this should fail because it is attempting to include two key columns in the projection.
* If the projection includes multiple copies of either `left.k` or `right.k`, then this should fail because it is attempting to include multiple key columns in the projection.
* If the projection includes `left.k` and `right.*`, then this should succeed with no copies of `right.k` in the resulting value columns, since the user did not explicitly request this with `AS_VALUE()`.
* If the projection includes `right.k` and `left.*`, then this should succeed with no copies of `left.k` in the resulting value columns, since the user did not explicitly request this with `AS_VALUE()`.
* If the right includes both ` no explicit key columns

Reply:
We can see from the unexpected behavior (unexpected in light ksqlDB's approach of requiring users to be strict about selected keys, selecting no more and no less, and using `AS_VALUE()` in order to explicitly move them to the value if desired) that there are many inconsistencies around how this is handled today. IMO discussing what the behavior should be (e.g., whether we want to lift the stringent requirements around key selection, not just for joins but for all queries) is a separate question from the immediate decision at hand which is what we want the behavior for FK joins to be, as we are introducing FK joins for the first time. I think that the existing inconsistencies that have been pointed out here should be considered bugs and that we should introduce FK joins with "proper" semantics and restrictions, in light of ksqlDB's overall approach today, even if that means we're inconsistent between the behavior of FK and PK joins. Then we can revamp all these semantics in the future as part of a larger project.

cc @mjsax @agavra @colinhicks who might have other opinions.

Reply:
On second thought, I have a hunch for where this odd behavior is coming from in the ksqlDB code base and suspect it's likely easier to propagate the existing behavior (where right key columns are magically copied into the join result value) for FK joins than to try to tighten up the semantics. If this is the case, I think we should go with this option since (1) it's actually much less weird for an FK join to copy right key columns into the result value, since otherwise the right key columns would not be represented in the result, and (2) for the sake of shipping FK joins sooner rather than later, since users who don't want right key columns in the result value can choose to leave them out via explicit selection.

I'll dig into the implementation a bit more and update this test suite to reflect the outcome.

Reply:
I was going to say that this odd behavior is not, all things considered, that odd. We can say that in a join the left key is always taken as the key, unless you select only the right key. I believe this was by design from Andy when he originally introduced this feature. We might be able to dig this up in the KLIP somewhere.

For FK joins, I'd recommend just sticking to the existing behavior as I don't think it's that important TBH to nail this. It might be a quirk, but as long as it can be clearly explained by one principle ("in any situation where you select both keys, the right gets copied into the value") it's not that urgent to fix.

Reply:
Thanks for the very nice summary @vcrfxia . So if we follow the direction to stick with the current semantics, I'd just want to clarify the following for added FK-join:

0. we would not allow `JOIN right ON left.k == right.v` but only `JOIN right ON left.v == right.k`. I.e. users need to make sure the FK is always on left-hand side.
1. If the projection includes `left.k` but not `right.k`, this should succeed with `left.k` as PK.
2. If the projection includes `right.k` but not `left.k`, this should fail.
3. If the projection includes both `left.k` and `right.k`, then this should fail; users need to move `right.k` to values via "as_value".
4. If the projection includes multiple copies of either left.k or right.k, then this should fail.
5. If the projection includes `left.k` and `right.*`, then this should succeed with `left.k` as PK and `right.k` in values automatically.
6. If the projection includes `right.k` and `left.*`, then this should fail; users need to move `right.k` to values via "as_value".
7. If the projection includes both `left.*` and `right.*` and no explicit key columns, then this should succeed with `left.k` as PK and `right.k` in values automatically.
8. If the projection contains a single (unqualified) * and no explicit key columns, then this should succeed with `left.k` as PK and `right.k` in values automatically.

Does that sounds right?

Reply:
Confirmed the suspicion above that consistent behavior with PK joins is what the current (ongoing) FK join implementation uses, so we'll go with this for the upcoming release. Updated this test suite to reflect the changes. This PR is now ready for final review and to be merged cc @mjsax .

Reply:
Thanks @vcrfxia , could you verify that the point.0 above is true as well ?

Reply:
@guozhangwang : 
0. This is correct. ksqlDB today is not set up to reverse the order of joins (for example, this is why only LEFT JOIN and not RIGHT JOIN is supported).
1. Correct.
2. Correct.
3. Incorrect, assuming that by "current semantics" you mean what ksqlDB currently does for PK joins which is to allow this and put `right.k` into the value by default. We've decided to mirror the existing behavior for PK joins for FK joins as well, so I will answer the rest of these questions assuming we're talking about this (though I still contend that the existing behavior for PK joins is buggy / inconsistent with how ksqlDB treats keys in projections for the non-join case 🙂)
4. The query will fail if there are multiple copies of `left.k`, but the query will not fail if there are multiple copies of `right.k` (assuming the different copies have been properly aliased to have different names, as ksqlDB does not allow multiple columns with the same name).
5. Correct.
6. Incorrect. The will succeed with `right.k` moved to the value automatically.
7. Correct. 
8. Correct.

Reply:
For 4/6, I'm stating that just based on the current added tests, where if user do not manually specify `as_value` it would fail. I'm fine with changing it to be aligned with the PK-join semantics, just clarifying that in this case we would need to change the tests here.

Reply:
Which tests are you referring to? I updated the PR before my previous replies, wondering if it's possible you were looking at an older version.

Reply:
Obviously I did not get the latest commit :) Sorry